docs/spec/broker.md

```diff
@@ -40,10 +40,41 @@ default Broker upon creation if no Broker is specified by the user.
 A Trigger MAY be created before its assigned Broker exists. A Trigger SHOULD
 progress to Ready when its assigned Broker exists and is Ready.
 
+#### Event filters
+
+Implementations of Trigger MUST support both attributes filter and jsExpression
```

Comment:
I would rather use Rego. https://www.openpolicyagent.org/docs/v0.11.0/language-reference/

It is designed to do this kind of filtering and they have a cross-compiler to web assembly that is stupid fast for filter functions.

Reply:
cc: @grantr

Reply:
> It is designed to do this kind of filtering and they have a cross-compiler to web assembly that is stupid fast for filter functions.

That's not easy as trivial as it seems, i've looked into it: you need a wasm engine, then you need to implement this ABI https://www.openpolicyagent.org/docs/latest/wasm/#exports, then you need to bring the event into the wasm context (which requires a serde step). JS is far easier to integrate and easier also for the users to look at it

Reply:
Then rego brings the non-trivial problem that you need to compile the expressions in wasm modules, invoking the rego compiler, and store them somewhere

Reply:
I don't think it is less trivial than embedding a javascript interpreter, and for this to be able to be well understood we need to define or point to a spec of everything you can do in a `jsExpression` similar to how Rego lists out all the language features.

Plus we don't need to implement rego, we pull OPA in as a sidecar as one impl.

Reply:
> I don't think it is less trivial than embedding a javascript interpreter

Look at the code, it's one file of 100 lines of code :smile: Any sidecar solution requires far more, both on sidecar and on filter code + all complexity of configs and tests too. And notice, this proposal doesn't prevent to have in future rego too :smile:

> we need to define or point to a spec of everything you can do in a jsExpression

Not really, I was wondering that all we should say here is: the js engine supports EcmaScript 5.1 (which is a requirement every js engine out there supports today) and should serve cloudevent as `event` (and i think i already covered that part). Maybe we should be more specific on the type coercion between cloudevents types and js types, but overall that's all i think we need here 

> Plus we don't need to implement rego, we pull OPA in as a sidecar as one impl.

That adds quite some overhead to the filtering process and complexity too... I prefer to have something simple that I can embed straight in my code, other solutions that requires some external process are more complex to implement and most notably adds overhead for serializing/deserializing and IPC.

Reply:
I'm not convinced on the "MUST support both attributes filter and jsExpression".

I actually like the filter dialects the CloudSubscriptions spec talks about. See https://github.com/cloudevents/spec/blob/master/subscriptions-api.md#323-filter-dialects. 
The MUST is to support a basic one (Exact, Prefix and Suffix match).
More advanced dialects can be made optional. I can easily see JS being an optional advanced dialect. Maybe REGO, CEL, or whatever, other ones, and so on...

Overall, I think you will be able to reduce contention if you decouple Trigger from a mandatory JS "dialect", and just frame the problem as Trigger with optional dialects... I think conversations will move forward faster, and we will also be more aligned with the Serverless WG (which in general I think is a good thing).

Reply:
TBH I wasn't aware of these filter dialects from subscriptions spec. Still they don't seem to cover basic nested logic and operations on types (in this doc is explained what i think a filtering expressions should let the user express https://docs.google.com/document/d/1Pz2vaLWKUrMQDLNyDW7ksjgLG4GAxBY546dT7_jF5Rk/edit?usp=sharing)

My take on optional and mandatory is that, I have no strong opinion if vendors wants to add their filtering capabilities to their system, but IMHO eventing vanilla should be bundled with a filtering capability that covers a 90% of basic use cases. Not providing a "good enough" bundled filtering capability will make too big the divergence between different implementations (basically, if a user wants to port trigger a to another vendor, he needs to rewrite all the filtering logic...)

Reply:
I would rather look at what is missing in the subscription spec and upstream changes and take what is there for now.

Reply:
crazy thought... what if you proposed a javascript dialect to the spec https://github.com/cloudevents/spec/blob/master/subscriptions-api.md#323-filter-dialects